proof/invariant-abstract/DetSchedInvs_AI.thy

```diff
@@ -584,23 +584,15 @@ end
 definition tcb_of :: "kernel_object \<rightharpoonup> tcb" where
   "tcb_of ko \<equiv> case ko of TCB tcb \<Rightarrow> Some tcb | _ \<Rightarrow> None"
 
-lemma tcb_of_simps:
-  "tcb_of (TCB tcb) = Some tcb"
-  "tcb_of (SchedContext sc n) = None"
-  "tcb_of (CNode n cnode) = None"
-  "tcb_of (Endpoint ep) = None"
-  "tcb_of (Notification ntfn) = None"
-  "tcb_of (Reply reply) = None"
-  "tcb_of (ArchObj aobj) = None"
-  by (auto simp: tcb_of_def)
-
-lemma tcb_of_Some:
+lemmas tcb_of_simps [simp] = tcb_of_def [split_simps kernel_object.split]
+
+lemma tcb_of_Some[simp]:
   "tcb_of ko = Some tcb \<longleftrightarrow> ko = TCB tcb"
```

Comment:
I don't think this one is right yet. What I meant was that we should have something like
```
(tcb_of ko = Some tcb) = (ko = TCB tcb)
```
in the simpset (which may already exist later, not sure). But we should not have `tcb_of_SomeD` in the simp set, because that would apply to any term of type `kernel_object` (it'll fail for most of them, but it will slow down all other proofs). If we have the rule above in the simpset, then `tcb_of_SomeD` is just that rule with `[THEN iffD1]`, which we should either prove that way if we actually need it as a dest rule, or leave out completely.

Reply:
This is iff and this is the version that you added?

Reply:
This is the pre-existing one and I didn't change the statement. I just added it to the simp set  (I had to change the proof, though).

Reply:
I can't find that rule, did I already add it? In any case, it should be added if it doesn't exist yet. If it's already there somewhere, it'd be good to move it close to the definition.

`[iff]` is probably safe for that one, yes, which means we can leave out the current form `tcb_of_Some` because that should be generated as part of `iff`.

Reply:
What I am saying is that the statement has "longleftrightarrow". It is an implication in both ways.

Reply:
Oh dear, my brain so did not see that, sorry, I kept seeing "longrightarrow". Thanks for pointing that out. And then everything is fine as is. We might make it `iff`, but it's not a priority.